shell/shared/grain.js

```diff
@@ -1148,52 +1312,68 @@ Router.map(function () {
       ];
     },
 
-    data: function() {
-      if (!this.ready || Meteor.loggingIn()) {
-        return;
-      }
-      if (Meteor.userId() && !Session.get("visit-token-" + this.params.token)) {
-        var tokenInfo = TokenInfo.findOne(this.params.token);
-        if (!tokenInfo || !tokenInfo.apiToken) {
-          this.state.set("error", "invalid authorization token");
-        } else {
-          this.state.set("error", undefined);
-          var apiToken = tokenInfo.apiToken;
-          if (!Grains.findOne({_id: apiToken.grainId, userId: Meteor.userId()}) &&
-              !ApiTokens.findOne({userId: apiToken.userId, "owner.user.userId": Meteor.userId()})) {
+    onRun: function () {
+      var token = this.params.token;
+      var path = this.params.path;
+      var query = this.params.query;
+      var hash = this.params.hash;
+
+      Tracker.autorun(function() {
+        var tokenInfo = TokenInfo.findOne({_id: token});
+        if (tokenInfo && tokenInfo.apiToken) {
+          console.log("valid");
+          var grainId = tokenInfo.apiToken.grainId;
+          if (!Grains.findOne({_id: grainId, userId: Meteor.userId()}) &&
+              !ApiTokens.findOne({userId: tokenInfo.userId, "owner.user.userId": Meteor.userId()})) {
             // The user neither owns the grain nor holds any sturdyrefs from this sharer.
             // Therefore, we ask whether they would like to go incognito.
             // TODO(soon): Base this decision on the contents of the Contacts collection.
-            return {interstitial: true, token: this.params.token};
+          }
+          var grains = globalGrains.get();
+          var grainIndex = grainIdToIndex(grains, grainId);
+          if (grainIndex == -1) {
+            var openView = function openView() {
+              var mainContentElement = document.querySelector("body>.main-content");
+              if (mainContentElement) {
+                var grains = globalGrains.get();
+                var grainToOpen = new GrainView(grainId, path, query, hash, token,
+                    mainContentElement);
+                grainToOpen.openSession();
+                grainIndex = grains.push(grainToOpen) - 1;
+                globalGrains.set(grains);
+                makeGrainIdActive(grainId);
+              } else {
+                Meteor.defer(openView);
+              }
+            };
+            openView();
           } else {
-            Session.set("visit-token-" + this.params.token, "redeem");
+            makeGrainIdActive(grainId);
           }
+
+          var currentGrain = getActiveGrain(grains);
+          if (!currentGrain.shouldShowInterstitial() && currentGrain.sessionStatus() == "closed") {
+            currentGrain.openSession();
```

Comment:
Perhaps `openSession()` should be renamed "tryOpenSession" - it will not actually open the session if `shouldShowInterstitial()` is true (because we need the user's input to decide whether incognito should be true or false in the newApiTokenSession).  To make sure that we attempt to open the session as soon as that value becomes false, we need this branch, since `openView()` is only called when we hadn't created a GrainView yet.

(Even better, we should guard the `openSession()` call in `openView()` with a check to `shouldShowInterstitial()`.  I think I had that check in the original code, and lost it in the openView rework after Kenton did the explicit Blaze rendering, which requires we be able to get a handle to the parent DOM node.)